### ISSUE TITLE
Handling `da.histogram` when `range` is not specified

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#4544]
Currently `da.histogram` is unable to handle the case where `range` is not specified or for that matter when `range` consists of lazy values. It would be nice for `da.histogram` to handle this case by simply using the min/max of the data as the default `range` (as is done in `np.histogram`). Further it would be good if `da.histogram` avoided (accidentally) computing lazy values used for the `range`.

**Comment by mrocklin** [comment]
I agree that this will be ideal.  This will also force all of the data to be in memory at once (need to compute min/max before we can compute any histogram).  Given that, do we still want this behavior, or should we inform the user about the limitation and let them make their own choice?

In practice, my guess is that some other algorithm like approximate quantiles is likely to be more useful for users.

**Comment by jakirkham** [comment]
Let's ask @GenevieveBuckley and @jni if they have thoughts 🙂

**Comment by GenevieveBuckley** [comment]
I agree that if `range` is left unspecified then dask should just work something sensible out, and that would be as close as possible to the behaviour of `np.histogram()`.

We could have the default be calculating the min/max and warn the user about the extra computation.

The fallback option the warning could suggest might be using a min & max based on the dtype of the input array. Again, the user should not have to fiddle around with it, but something like the way scikit-image [does exactly this](https://github.com/scikit-image/scikit-image/blob/master/skimage/exposure/exposure.py) is reasonably user friendly. The drawback is that you can get back histograms that are half zero if the dtype is signed but the actual data is not.

I do *not* like the idea of making three passes over the data in the worst case scenario (one for min, one for max, one for the actual histogram). It seems like it should certainly be possible to simultaneously compute the min AND max in a single pass over the data. Does this already exist somewhere in dask?

@mrocklin maybe I misunderstand your comment about approximate quantiles. In my opinion, that would be a very useful function for have for arrays but I see that as fulfilling a different need than a histogram function. For me:
histogram = used most often for finding thresholds for segmenting objects within an image (generally needs more detail than quartile buckets, 256 bins is common)
approximate quartiles = aggregation of statistics after the bulk of my analysis

**Comment by jni** [comment]
@GenevieveBuckley 

> It seems like it should certainly be possible to simultaneously compute the min AND max in a single pass over the data. Does this already exist somewhere in dask?

I had a nice chuckle reading this because @jakirkham ranted to me about this issue just last month. =D

https://github.com/numpy/numpy/issues/9836

> maybe I misunderstand your comment about approximate quantiles

I think you did. =) Firstly, quantiles and quartiles are different things. Secondly, the idea is that the distribution of the data should not change if you randomly subsample the data unless you are really unlucky. Therefore, to compute decent histogram bins (or even a decent histogram!) you only need to read, say, 10,000 points from your array (rather than all 100TB or whatever).

Unless *I'm* misunderstanding. =P

**Comment by GenevieveBuckley** [comment]
Ah, thank you for clarifying @jni. So, more along the lines of this?

Manku, *et al.* 1998 ['Approximate Medians and other Quantiles in One Pass and with Limited Memory'](https://dl.acm.org/citation.cfm?id=276342)
(PDF download [here](http://www.cs.umd.edu/~samir/498/manku.pdf))

Chen, *et al.* 2018 ['Quantile Regression Under Memory Constraint'](https://arxiv.org/abs/1810.08264)
(PDF download [here](https://arxiv.org/pdf/1810.08264.pdf))

**Comment by GenevieveBuckley** [comment]
This issue is causing some failing tests when CellProfiler experiments with using Dask arrays (for a specific example, see `TestAlign` failing in [cellprofiler/core](https://github.com/cellprofiler/core) on the `daskv2` branch).

cc @bethac07 (you might like to follow this tracking issue)
